crates/payload/validator/src/cancun.rs

```diff
@@ -0,0 +1,122 @@
+
+/// Checks block and sidecar w.r.t new Cancun fields and new transaction type EIP-4844.
+///
+/// Checks that:
+/// - Cancun fields are present if Cancun is active and vv
+/// - contains EIP-4844 transactions if Cancun is active and vv
+/// - checks blob versioned hashes in block and sidecar match
+#[inline]
+pub fn ensure_well_formed_fields<T, B>(
+    block: &SealedBlock<B>,
+    cancun_sidecar_fields: Option<&CancunPayloadFields>,
+    is_cancun_active: bool,
+) -> Result<(), PayloadError>
+where
+    T: Transaction + Typed2718,
+    B: Block<Body = BlockBody<T>>,
+{
+    ensure_well_formed_header_and_sidecar_fields(block, cancun_sidecar_fields, is_cancun_active)?;
+    ensure_well_formed_transactions_field_with_sidecar(
+        block.body(),
+        cancun_sidecar_fields,
+        is_cancun_active,
+    )
+}
+
+/// Checks that Cancun fields on block header and sidecar are present if Cancun is active and vv.
+#[inline]
+pub fn ensure_well_formed_header_and_sidecar_fields<T: Block>(
+    block: &SealedBlock<T>,
+    cancun_sidecar_fields: Option<&CancunPayloadFields>,
+    is_cancun_active: bool,
+) -> Result<(), PayloadError> {
+    if is_cancun_active {
+        if block.blob_gas_used().is_none() {
+            // cancun active but blob gas used not present
+            return Err(PayloadError::PostCancunBlockWithoutBlobGasUsed)
+        }
+        if block.excess_blob_gas().is_none() {
+            // cancun active but excess blob gas not present
+            return Err(PayloadError::PostCancunBlockWithoutExcessBlobGas)
+        }
+        if cancun_sidecar_fields.is_none() {
+            // cancun active but cancun fields not present
+            return Err(PayloadError::PostCancunWithoutCancunFields)
+        }
+    } else {
+        if block.blob_gas_used().is_some() {
+            // cancun not active but blob gas used present
+            return Err(PayloadError::PreCancunBlockWithBlobGasUsed)
+        }
+        if block.excess_blob_gas().is_some() {
+            // cancun not active but excess blob gas present
+            return Err(PayloadError::PreCancunBlockWithExcessBlobGas)
+        }
+        if cancun_sidecar_fields.is_some() {
+            // cancun not active but cancun fields present
+            return Err(PayloadError::PreCancunWithCancunFields)
+        }
+    }
+
+    Ok(())
+}
+
+/// Checks transactions field and sidecar w.r.t new Cancun fields and new transaction type EIP-4844.
+///
+/// Checks that:
+/// - contains EIP-4844 transactions if Cancun is active and vv
+/// - checks blob versioned hashes in block and sidecar match
+#[inline]
+pub fn ensure_well_formed_transactions_field_with_sidecar<T: Transaction + Typed2718>(
+    block_body: &BlockBody<T>,
+    cancun_sidecar_fields: Option<&CancunPayloadFields>,
+    is_cancun_active: bool,
+) -> Result<(), PayloadError> {
+    if is_cancun_active {
+        if !block_body.has_eip4844_transactions() {
```

Comment:
this check is wrong I believe

Reply:
should the error variant be removed from alloy?

Reply:
https://github.com/alloy-rs/alloy/issues/2081

Reply:
I think so, i believe this was intended for the sidecar fields which we check seperately